### ISSUE TITLE
@require and @resource download hash-check

### ISSUE THREAD

**Issue by pyhedgehog** [greasemonkey/greasemonkey#2349]
After reading #2331 and #1473 I have a note about security and trust:
User should trust script author not to download malicious resources, but author has no control on user environment (e.g. invalid trusted certificates).
To improve script consistency I propose to add ability to specify hash in addition to resource's URL.
There are semi-standard notation used by many download managers: "`http://www.example.com/logo.png#md5=1234...`", "`http://www.example.com/logo.png#sha256=1234...`" and so on.

**Comment by the8472** [comment]
The hash seems kinda hacky. New @ directives might be cleaner.

For `@resource` it's trivial. just add `@resourceIntegrity resourceName sha1 hash`. `@require` is a little more tricky, but if we stipulate that either none or all dependencies must be hashed then simply specifying a list of hashes should be enough and GM could then just check script hashes against the hash list.

**Comment by pyhedgehog** [comment]
I've thought about it, but it will be harder to make compatible with other userscript implementations (opera, tampermonkey and so on). Also this "hacky" decision is compatible with already existent software pieces as I mentioned earlier.

From my point of view checking all requires against all hashes is much more "hacky" and less secure (faking any of several hashes is N times easier that only one).

**Comment by the8472** [comment]
> but it will be harder to make compatible with other userscript implementations (opera, tampermonkey and so on).

how so? directives that are not understood will just be ignored

> From my point of view checking all requires against all hashes is much more "hacky" and less secure (faking any of several hashes is N times easier that only one).

having a separate hash directive makes it easier to specify multiple hashes per file for forwards-compatibility. e.g. if sha2 were to become too weak browsers that support sha3 it could check sha3 and those which don't still could check sha2.

a little more tedious option would be to have the `@requireIntegrity` directive spell out the URL a second time.

`@requireInegrity <url> <require-name> <algo=hash,algo=hash,...>`

would have the added benefit of associating a name with the require which could then be used in future directives, just like resources

**Comment by tiansh** [comment]
Why not just use `https` to make it secure?
Using `https` do not need any modification on GM.
Maybe add a requirement that @ require URI must be `https` by default makes more sense.

BTW, MD5 checksum do not make it more secure due to collision attacks.

**Comment by the8472** [comment]
https only verifies the connection, not that the server sends the correct content intended by the script author. This is especially important when fetching dependencies from 3rd party sources, e.g. CDNs.

**Comment by tiansh** [comment]
@the8472 Yes, `HTTPS` only makes sure the connection is secure. Then, script author should host the file on their own server or some servers which can be trusted.

**Comment by the8472** [comment]
Verification is much easier than ensuring trustworthiness indefinitely into the future.

CDNs / 3rd party includes also are a reality which you cannot really ignore.

**Comment by SINE** [comment]
Nice suggestion, thanks @arantius for putting that into consideration.

**Comment by JamesTheAwesomeDude** [comment]
> The url fragment seems kinda hacky

While I agree, it's apparently also what PyPI uses...

> script author should host the file on their own server or some servers which can be trusted.

This is a nice idea in theory, but very few people run their own server, and CDNJS is _so convenient_ to use and is arguably more clear than script-author-hosted imports as to what exactly is being included. (I'd be interested to see some statistics on CDN usage from, say, the admin of Greasyfork.)

> MD5 checksum do not make it more secure due to collision attacks.

Setting aside the fact that these collision attacks generally require an attacker-crafted library in the first place ([no useful general preimage attack yet exists for MD5](https://crypto.stackexchange.com/a/41865)), this is trivially fixed by the script author choosing to instead use SHA-256 for subresource integrity; an attacker in a position to sub a weaker algorithm in could just as easily sub in malicious code in directly, so it's somewhat moot from an implementation standpoint anyway.

**Comment by sjehuda** [comment]
I'm for the proposal of @the8472 +1
